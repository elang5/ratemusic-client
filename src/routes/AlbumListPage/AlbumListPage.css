.album-list-page {
  padding: .5rem;
  background: #1f1f21;
}

.album-list {
  padding: 0;
  list-style-type: none;
  display: grid;
  grid-template-columns: repeat(auto-fit, minmax(320px, 1fr));
}

.container {
  text-align: center;
}

@media only screen and (max-width: 720px) {
  .album-list-page {
    display: grid;
    justify-content: center;
  }
}