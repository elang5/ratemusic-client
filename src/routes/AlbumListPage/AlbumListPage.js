import React, { Component } from 'react'
import AlbumApiService from '../../services/albums-api-service'
import AlbumItem from '../../components/AlbumItem/AlbumItem'
import AlbumsApiService from '../../services/albums-api-service';
import SearchForm from '../../components/SearchForm/SearchForm';
import './AlbumListPage.css'

export class AlbumListPage extends Component {
  state = {
    albums: [],
    error: null,
  }

  componentDidMount() {
    AlbumApiService.getAlbums()
      .then(res => this.setState({
        albums: [...res]
      }))
      .catch(res => this.setState({
        error: res.error
      }))
  }

  renderAlbums = () => {
    const { albums } = this.state
    return albums.map(album =>
      <li className="album">
        <AlbumItem
          key={album.id}
          album={album}
          rating={this.getAverageRatings(album.id)}
        />
      </li> 
      )
  }

  searchAlbums = (album) => {
    const { albums } = this.state
    const matchedAlbums = albums.filter(albumName => {
      return albumName.title.includes(album)
    })
    this.setState({
      albums: matchedAlbums
    })
  }
  
  getAverageRatings = (albumId) => {
    const { albums } = this.state
    const averageRating = AlbumsApiService.getAlbumReviews(albumId)
        .then(res => res.reduce((sum, review) => {
          return (sum + review.rating)
        }, 0) / res.length)
        .then(value => parseFloat(value).toFixed(1))
        .then(rating => albums[albumId - 1].rating = rating)
        .catch(err => this.setState({ error: err.error }))
    return averageRating
  }

  render() {
    const { error } = this.state
    return (
      <>
        <SearchForm name={'Search for Albums: '} getAlbums={this.searchAlbums} />
        <section className="album-list-page">
          {error && <p className="error">There was an error. Please try again.</p>}
          <ul className="album-list">
            {this.renderAlbums()}
          </ul>
        </section>
      </>
    )
  }
}

export default AlbumListPage
