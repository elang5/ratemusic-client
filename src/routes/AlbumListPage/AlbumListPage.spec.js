import React from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'
import AlbumListPage from './AlbumListPage'

describe(`AlbumListPage Component`, () => {

  it(`renders the AlbumListPage component`, () => {
    const wrapper = shallow(<AlbumListPage />)
    expect(toJson(wrapper)).toMatchSnapshot()
  })
})