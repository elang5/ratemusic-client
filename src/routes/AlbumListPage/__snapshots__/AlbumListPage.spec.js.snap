// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AlbumListPage Component renders the AlbumListPage component 1`] = `
<Fragment>
  <SearchForm
    albums={Array []}
    handleSubmit={[Function]}
    name="Search for Albums: "
    searchAlbums={[Function]}
  />
  <section
    className="album-list-page"
  >
    <Loader
      loading={true}
    />
    <div
      className="container"
    >
      <ul
        className="album-list"
      />
    </div>
  </section>
</Fragment>
`;
